File: leetcodeJava/64.java
package leetcodeJava;

class Solution {
    public int minPathSum(int[][] grid) {
        int m = grid.length, n = grid[0].length;
        for (var i = 0; i < m; i++) {
            for (var j = 0; j < n; j++) {
                if (i == 0) {
                    if (j > 0)
                        grid[i][j] += grid[i][j-1];
                } else if (j == 0) {
                    grid[i][j] += grid[i-1][j];
                } else {
                    grid[i][j] += Math.min(grid[i][j-1], grid[i-1][j]);
                }
            }
        }
        return grid[m-1][n-1];
    }
}

File: scala/64.scala
object Solution {
    def minPathSum(grid: Array[Array[Int]]): Int = {
        for (i <- 0 until grid.length; j <- 0 until grid.head.length) {
            (i, j) match {
                case (0, 0) => ()
                case (0, _) => grid(0)(j) = grid(0)(j-1) + grid(0)(j)
                case (_, 0) => grid(i)(0) = grid(i-1)(0) + grid(i)(0)
                case (_, _) => grid(i)(j) = math.min(grid(i-1)(j), grid(i)(j-1)) + grid(i)(j)
            }
        }
        grid.last.last
    }
}